### PR TITLE
Reschedule/Re-position/Reset multiple cards at once

### LINKED ISSUE #2076: Reset Deck
Originally reported on Google Code with ID 1174

```
I read all the reasons why you shouldn't reset a deck but it's simply the decision of
the student if he wants to waste his time learning stuff he might not have forgotten
yet.

That said let me state my situation:
I studied Japanese 2 years ago and my deck had 400 in it. I recently decided to get
back to it, but now have all the cards for review while I know almost none of them
anymore PLUS if I DO know one it schedules the next review for 1 year into the future,
even if it was a lucky guess.

This is discouraging and the reason why I've been wasting my time looking for a solution
instead of learning.

Please add this feature; people will appreciate and those who don't just won't use
it.
```

Reported by `Hashishin` on 2012-06-16 10:34:49

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description
On of our[ most popular pieces of content](https://github.com/ankidroid/Anki-Android/graphs/traffic) in the whole project is the issue tracking resetting deck progress (either making cards new, or completely resetting review status).

The infrastructure for actually doing this was nailed by @vrublack's excellent card browser multi-select PR #4839

This aims to extend multi-select in the card browser by adding three new action menu items - to reschedule, re-position, or reset multiple cards in the same way NoteEditor does it for a single card

It also implements undo for both the original single-edit case as well as multi-select while listening for memory pressure

## Fixes
Fixes #2076 

## Approach
- First commit refactored DeckTask.TaskListener to remove boilerplate
- Second commit refactored reschedule and reposition on NoteEditor to do the work with DeckTask where it was on the UI thread before
- Third commit refactored reset on NoteEditor to also use DeckTask

At that point there's no new feature yet, just the infrastructure for the extra functionality

- Fourth commit - added the feature in CardBrowser as an analog to the feature in NoteEditor but on multiple cards
- Fifth commit implements undo for both the single and multi case

- Sixth commit adopts the change from #5119 into this framework
- Seventh commit responds to request to move all single-card actions from NoteEditor to Reviewer

- Eighth commit is non-semantic codacy / lint changes

## How Has This Been Tested?

So far just existing tests and manual tests on emulator. If I had the SQLite/Robolectric action from #5103 I'd have coverage here...

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
@timrae  with apologies as I said I would do the template editor first but once I saw this was doable in an afternoon (more or less) I had to give it a try. This and #5103 is probably enough to chew on for review that I can work the template editor while you're digesting...maybe with Undo now that I've seen more of that sub-system in this PR

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/IntegerDialog.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        public void setInt(int intArg) {
+            mInt = intArg;
+        }
+        public int getInt() {
+            return mInt;
+        }
+        public abstract void run();
+    }
+
+    private IntRunnable callbackRunnable;
```
[the remark this answers is not in the document]
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/IntegerDialog.java]: :man_facepalming:

**Comment by mikehardy** [comment]
Unless there is a better way unknown to me, I believe there is no non-verbose solution to the "string proliferation for plurals" issue, and altering the strings to actually be be plurals (https://developer.android.com/guide/topics/resources/string-resource#Plurals) is the necessary conversion

**Comment by eginhard** [comment]
Plurals are tricky because other languages might not make just a binary distinction and can e.g. also have a separate form for 2 items (or any number ending in 2). But if you specify them like [the existing plural strings](https://github.com/ankidroid/Anki-Android/blob/cd0e1817a51e009527a55816fe3dc95e9401e817/AnkiDroid/src/main/res/values/01-core.xml#L40), it's straightforward for translators. CrowdIn will ask for all required forms in a language ([4 in Lithuanian](https://github.com/ankidroid/Anki-Android/blob/cd0e1817a51e009527a55816fe3dc95e9401e817/AnkiDroid/src/main/res/values-lt/01-core.xml#L59)!), pre-fill the alternatives with the singular form so it just needs a quick edit and then Android automatically shows the correct form based on the number and the language's grammar. _I think_ this just means that you'd need to change the strings to include a %d.

**Comment by mikehardy** [comment]
Thank @eginhard that was what I thought by reading but I haven't done it before so good to hear my read might be right

**Comment by mikehardy** [comment]
So the most serious question I have outstanding to make this a high quality patch is the undo functionality. Recognize that undo wasn't possible before anyway, but I'd like to do it in order to learn how it works. I think because the scheduling information is all within the card object, I need to clone the Java objects to preserve the information and then on undo I just one-by-one flush them? I'll verify this myself either way but similar to the plurals issue it's wonderful to have second opinions...

**Comment by mikehardy** [comment]
I just re-pushed, with a mix of android plurals usage where unavoidable, and simply not mentioning quantities where possible.

I also collapsed the single/multi case into multi since single is just one type of multi, really

All that is left is undo implementation but this is not a regression. At this point it's ready for review on the base implementation, while I investigate how undo works and see if I can add it

**Comment by mikehardy** [comment]
Leaving this note here for myself: as a technical difficulty for saving the cards for later undo, I need to deal with Java being a shallow-clone language for array copies, while the information I want to save is deep in the Card objects. There are a variety of ways to handle this but I think the Cloneable/protected-copy-constructor style here would fit us well: http://www.agiledeveloper.com/articles/cloning072002.htm - the only lingering doubt I have is about memory pressure as I'll be deep-copying an unbounded set of objects here. Even if I optimize for space by taking primitives of just the scheduling information I'm altering (which would be a maybe brittle but otherwise functional implementation). I'll either need to make a handler for memory pressure and make this cancellable because there could still be user on a small-ram device that tries to reschedule a huge collection.

So memory pressure callbacks and cancellability on the async task are important

**Comment by mikehardy** [comment]
Still leaving notes for myself, but the research and really understanding how to do things correctly is important to me. So the last step is to make sure that if I implement undo on an unbounded collection I can do so without causing an out of memory condition in the application.

It appears the right way to go about this is to [listen to the onTrimMemory()](https://developer.android.com/reference/android/content/ComponentCallbacks2.html#onTrimMemory(int)) calls from the system. There are ways to poll for memory status but they are either not recommended in lower APIs or [not available until API16](https://developer.android.com/reference/android/app/ActivityManager#getMyMemoryState(android.app.ActivityManager.RunningAppProcessInfo) ) so onTrimMemory in the application or activity is the correct approach for our app

Then connect that to the AsyncTask that implements the undo by calling cancel() on it and checking [isCancelled()](https://developer.android.com/reference/android/os/AsyncTask#cancelling-a-task) periodically from doInBackground()

I don't believe we have the infrastructure right now to connect our AsyncTask infrastructure with anything that gets the onTrimMemory calls, so that's the next item to figure out.

**Comment by mikehardy** [comment]
I've implemented undo for reset/reposition/reschedule activities both for single cards and multiple cards, while being sensitive to memory pressure per the research notes above

@timrae this one is good to go I think? You always find something though :-)

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java [hunk 2]
```diff
@@ -895,6 +890,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 startActivityForResultWithAnimation(intent, REQUEST_ADD, ActivityTransitionAnimation.LEFT);
                 return true;
             }
+            // FIXME convert these to Async - Dialog box for input/confirm, DeckTask implementation, listener here
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: handled fully in later commits

**Comment by mikehardy** [comment]
@timrae I just adopted the change from #5119 into the async/undo style here, so this is complete and ready to go I think (with the same caveats as above - looking at each commit individually is usually best, I try to package them so they are coherent one by one and build on each other)

**Comment by timrae** [comment]
Thinking about it properly, it seems pretty wasteful to have reset functionality in 3 places. I think it makes sense to remove it from the note editor and make it enabled by default in the reviewer... The only path to the note editor is from the reviewer or the browser. What do you think?

**Comment by mikehardy** [comment]
I don't like the implementation copy either but is the goal to reduce copying or mirror where available in desktop or is this an android UX preference? I'll check desktop when I have a computer near me again

**Comment by timrae** [comment]
I believe desktop only provides the feature in the browser... It certainly doesn't give it in three separate places

**Comment by mikehardy** [comment]
Thinking more, yes. Now that multi of all forms works in the browser, remove from editor and make it an overflow in reviewer. Maybe all 3 in reviewer menu (reschedule reposition reset) ? Seemed easy to do the svg thing there and the code is a simple move? Or just the 3 in browser and reset button in review and done?

**Comment by mikehardy** [comment]
I like the idea of the 3 as singles in review fwiw, very convenient vs use case of going to browser and searching to do them

**Comment by timrae** [comment]
3 sounds a bit overwhelming, considering desktop doesn't offer this functionality AFAIK.  How about collapsing them into a single item like we do with suspend and bury?

**Comment by mikehardy** [comment]
I would use the "history" icon for reschedule since it references reset but with the idea of time in it https://material.io/tools/icons/?icon=history&style=baseline

And maybe this one for the same reference (they all have the same circle-arrow thing going on) but with the dot inside to indicate a specific spot for reposition: https://material.io/tools/icons/?icon=settings_backup_restore&style=baseline

**Comment by timrae** [comment]
Suspend and bury each have a submenu for note vs card

**Comment by mikehardy** [comment]
💡 

I'd never used them! Okay, can definitely do that.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java [hunk 3]
```diff
@@ -171,148 +172,220 @@ public boolean commit() {
...
-                            mOptions.getJSONObject("new").put("initialFactor", (Integer) value * 10);
-                        } else if (key.equals("newOrder")) {
-                            int newValue = Integer.parseInt((String) value);
-                            // Sorting is slow, so only do it if we change order
-                            int oldValue = mOptions.getJSONObject("new").getInt("order");
-                            if (oldValue != newValue) {
-                                mOptions.getJSONObject("new").put("order", newValue);
-                                DeckTask.launchDeckTask(DeckTask.TASK_TYPE_REORDER, mConfChangeHandler,
-                                        new DeckTask.TaskData(new Object[] { mOptions }));
+                        switch (key) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java]: AndroidStudio auto-convert to switch, then following convention case statements that had variable assignments got extra curly braces for scope

**Comment by mikehardy** [comment]
@timrae I think this is re-re-ready with the changes requested, and I re-based and re-pushed

Looks like this: https://youtu.be/DyNF3KB9F5o

**Comment by timrae** [comment]
Definitely don't put reschedule at the top of the menu...

**Comment by mikehardy** [comment]
Is there a convention for where to put menu items? I'm not seeing it as alphabetical when I click around, but I don't care where it goes. I'll drop it to the bottom for now

No mention of how cool my first svg icon ever is even though I had to hand-craft the SVG path for that little drop-down triangle? That thing is ace! haha

**Comment by mikehardy** [comment]
@timrae re-pushed, scheduling sub-menu at the bottom

**Comment by mikehardy** [comment]
@timrae on the "ping for review" front, this is the coolest (IMHO) but most complicated review. I think it's ready though and the commits one-by-one aren't too bad to look through

**Comment by timrae** [comment]
Yeah honestly I don't have time to give this a proper review now, I have skimmed over it a couple of times, but haven't gone through and figured out the logic yet. Realistically I don't think I can get to this anytime soon, so maybe it's more realistic to just merge and fix any issues that come up later

**Comment by mikehardy** [comment]
>  maybe it's more realistic to just merge and fix any issues that come up later

fair enough - I can say I did my best for what that's worth - and I'll have the broom and dustpan ready for breakage. Thanks Tim

**Pull request merged**